src/Parsers/ASTColumnDeclaration.cpp

```diff
@@ -130,12 +130,25 @@ void ASTColumnDeclaration::formatImpl(const FormatSettings & format_settings, Fo
 
 void ASTColumnDeclaration::forEachPointerToChild(std::function<void(void **)> f)
 {
-    f(reinterpret_cast<void **>(&default_expression));
-    f(reinterpret_cast<void **>(&comment));
-    f(reinterpret_cast<void **>(&codec));
-    f(reinterpret_cast<void **>(&statistics_desc));
-    f(reinterpret_cast<void **>(&ttl));
-    f(reinterpret_cast<void **>(&collation));
-    f(reinterpret_cast<void **>(&settings));
+    auto visit_child = [&f](ASTPtr & member)
+    {
+        IAST * new_member_ptr = member.get();
+        f(reinterpret_cast<void **>(&new_member_ptr));
+        if (new_member_ptr != member.get())
+        {
+            if (new_member_ptr)
+                member = new_member_ptr->ptr();
+            else
+                member.reset();
+        }
+    };
+
+    visit_child(default_expression);
```

Comment:
Should we visit each child in such a way in all asts or only in `ASTColumnDeclaration` and `ASTTableJoin` ?

Reply:
This is only necessary when the members are shared pointers (vs naked pointers).

I introduced the old code in `ASTColumnDeclaration` in [a previous PR ](https://github.com/ClickHouse/ClickHouse/pull/71541) and AFAICT it did not cause any issues because the pointer was not modified again, but with `ASTTableJoin` it happened, which is why the issue appeared. `ASTViewTargets` already had this code from before. It will all go away once we remove the old analyzer, which is a huge pain